tests/BootstrapperTest.php

```diff
@@ -326,6 +327,38 @@
     expect(File::isDirectory($tenantStoragePath))->toBeFalse();
 });
 
+test('MailTenancyBootstrapper maps tenant mail credentials to config as specified in the $credentialsMap property', function() {
+    MailTenancyBootstrapper::$credentialsMap = [
+        'mail.mailers.smtp.username' => 'smtp_username',
+        'mail.mailers.smtp.password' => 'smtp_password'
+    ];
+
+    config([
+        'mail.mailers.smtp.username' => $defaultUsername = 'default username',
+        'mail.mailers.smtp.password' => 'no password'
+    ]);
+
+    Tenant::create(['smtp_password' => 'testing password'])->run(function() use ($defaultUsername) {
+        expect(array_key_exists('smtp_password', tenant()->getAttributes()))->toBeTrue();
+        expect(array_key_exists('smtp_host', tenant()->getAttributes()))->toBeFalse();
+        expect(config('mail.mailers.smtp.username'))->toBe($defaultUsername);
+        expect(config('mail.mailers.smtp.password'))->toBe(tenant()->smtp_password);
+    });
+});
+
+test('MailTenancyBootstrapper reverts the config to default when tenancy ends', function() {
+    MailTenancyBootstrapper::$credentialsMap = ['mail.mailers.smtp.password' => 'smtp_password'];
+    config(['mail.mailers.smtp.password' => $defaultPassword = 'no password']);
+
+    tenancy()->initialize(Tenant::create(['smtp_password' => $tenantPassword = 'testing password']));
+
+    expect(config('mail.mailers.smtp.password'))->toBe($tenantPassword);
+
+    tenancy()->end();
+
+    expect(config('mail.mailers.smtp.password'))->toBe($defaultPassword);
```

Comment:
Do you think we want to use `invade()` in this these tests as well? To ensure that it's not reflected just in the config, but also on the mailer?

Reply:
Yeah I think that makes sense

Reply:
Though then we should be clear about the difference between these tests and the tests in the newly added test file.

Reply:
I think BootstrapperTest could test that using the MailTenancyBootstrapper makes the key things the bootstrapper provides work (mapping the tenant attributes to config and reverting the config when tenancy ends, making the mailer use the tenant credentials, and making `app(MailManager::class)` resolve TenancyMailManager).

So because the current MailTest tests only test the core things that MailTenancyBootstrapper provides, we could move the tests to BootstrapperTest. We could also keep the MailTest file to test the more under-the-hood things, like ensuring that TenancyMailManager re-resolves the cached mailers correctly.

Reply:
I guess having some duplication is fine if the version in the MailTest file is more in-depth. Meaning, some tests can be similar with the BootstrapperTest version only asserting that the config keys & mailer properties are correct. And the MailTest verifying all the low-level details. And possibly that reverting (or switching multiple times between tenant & central, or between different tenants directly) works well.

So the BootstrapperTest could be simple like the ones that are in the file, just checking *that it works*, and the MailTest would check all of the details of how it works if that makes sense.

Reply:
I was testing the re-resolving and realized we actually don't even need that in the end.

I think the `get()` method overriding was primarily there to ensure the mailer credentials got updated in real-time (when the config and tenant credentials got updated).

Later, I learned that because the config updates only when tenancy gets bootstrapped, re-resolving should have no real effect, but because removing the `get()` method override would make the central app's mailer have the password of the previous tenant if tenancy got canceled using `tenancy()->end()`, I kept the override.

Now, I tried removing the override again, and I realized that after ending tenancy, the tenant credentials persist because the `TenancyMailManager` instance (which stores the mailers and their credentials) we bind in the mail bootstrapper stays bound to `MailManager::class` even after ending tenancy.

I suggest deleting the `TenancyMailManager` class and updating `MailTenancyBootstrapper` so that it just binds a new `MailManager` instance to `MailManager::class` (instead of new `TenancyMailManager`), and so that it does the same thing in `revert()` (`app()->forgetInstance(MailManager::class)` didn't work, but extending `MailManager::class` the same way as in `bootstrap()` did).

Reply:
I also updated the tests

Reply:
So is everything here done now and ready for review?

One question about the implementation: couldn't we just use `forgetInstance()` in both places, like we use `bindNewMailManagerInstance()` now?

Using `extend()` repetitively seems like a bad way to implement this, I think each `extend()` registers a new "extension", so the closure will run several times, and also we're kind of abusing `extend()` here.

Getting rid of the override is definitely good, though I'd look for a better solution than registering `extend()` Closures an unlimited number of times for the same binding.

Reply:
I realized I could try forgetting `mail.manager` instead of `MailManager::class`, and that worked. So I got rid of `bindNewMailManagerInstance()` and used `forgetInstance()` instead.

> So is everything here done now and ready for review?

Yeah